ml_pipeline_engine/dag/manager.py

```diff
@@ -526,27 +545,46 @@ async def _run_oneof(self, node_id: NodeId) -> t.Any:
 
         logger.debug('Prepare OneOf DAG node_id=%s', node_id)
 
-        for subgraph_node_id in self.dag.graph.nodes[node_id][NodeField.oneof_nodes]:
-            logger.debug(
-                'Prepare OneOf subgraph to start target_node_id=%s, subgraph_node_id=%s',
-                node_id,
+        for idx, subgraph_node_id in enumerate(self.dag.graph.nodes[node_id][NodeField.oneof_nodes]):
+
+            oneof_dag = self._get_reduced_dag_input_one_of(
+                source=self.dag.input_node,
+                dest=subgraph_node_id,
+            )
+
+            logger.debug('Prepare [%s]%s to start. OneOf result node %s', idx, oneof_dag, node_id)
+
+            self._create_task(coro=self._run_dag(dag=oneof_dag), name=str(oneof_dag))
+
+            await self._lock_manager.wait_for_condition(
                 subgraph_node_id,
+                lambda: (
+                    self.__has_subgraph_error(oneof_dag)  # noqa: B023
+                    or self._node_storage.exists_result_type(
+                        subgraph_node_id,  # noqa: B023
+                        exclude_type=(Recurrent,),
+                    )
+                ),
             )
 
-            with suppress(OneOfSubgraphDagError):
+            if not self.__has_subgraph_error(oneof_dag):
 
-                return await self._run_dag(
-                    dag_output_node=node_id,
-                    is_under_oneof=True,
-                    dag=self._get_reduced_dag_input_one_of(
-                        self.dag.input_node,
-                        subgraph_node_id,
-                    ),
-                )
+                # The node_id is a synthetic node and cannot be executed anywhere. Hence, we should copy the
+                # result of the last successful subgraph and unlock everything related to the synthetic node.
+                self._node_storage.copy_node_result(subgraph_node_id, node_id)
 
-        raise OneOfDoesNotHaveResultError(node_id)
+                await self.__unlock_itself(node_id)
+                await self.__unlock_descendants(node_id)
+                await self.__unlock_run_method()
+
+                logger.debug('The %s has been succeeded', oneof_dag)
+                return
+
+        await self.__raise_exc(
+            OneOfDoesNotHaveResultError(node_id),
+        )
 
-    async def _run_switch(self, node_id: NodeId, is_under_oneof: bool) -> t.Any:
+    async def _run_switch(self, dag: DiGraph, node_id: NodeId) -> t.Any:
```

Comment:
I do like that we started to pass the dag object in the methods method; 
and am I correct that we cannot save this particular dag that was passed in the method to the "self" of class because it can be call of a subgraph (which is not equal to the original graph) and it is call with recursion ?

Reply:
Yes